### ISSUE TITLE
ESNI lifetime implications from Alt-Svc vs HTTPSSVC precedence

### ISSUE THREAD

**Issue by davidben** [MikeBishop/dns-alt-svc#105]
The draft originally said (prior to #66) said the Alt-Svc cache overrides HTTPSSVC. This had problems (issue #58 and #60), so #66 downgraded it to a MAY. It seems this still has problems.

**HTTPSSVC vs Alt-Svc**

Consider a server which uses both QUIC and ESNI. It configures both in HTTPSSVC. It also cares about HTTPSSVC-less clients (older client or legacy DNS resolver), so it configures QUIC in Alt-Svc. Is it required to configure ESNI in Alt-Svc, or can it leave things alone (with the understanding that ESNI will be limited to clients that support HTTPSSVC)?

The spec currently says a client (that would otherwise support HTTPSSVC) MAY skip an HTTPSSVC lookup given an Alt-Svc cache entry. That means, for ESNI to work, the server MUST configure it in both. This is not obvious and should be written down. More importantly, it has deployment consequences.

HTTPSSVC records apply to the *current* HTTP request. If the client has no cached DNS record, it still queries DNS and gets HTTPSSVC. That means HTTPSSVC TTLs may be set more-or-less freely depending on the site's performance vs. flexibility needs. Let's say it's O(1 hour).

Alt-Svc headers apply to *subsequent* HTTP requests. If the client has no Alt-Svc entry cached, it will send the HTTP request without Alt-Svc. That means Alt-Svc TTLs must cover the time to the next HTTP request for Alt-Svc to be used at all. For reference, I see google.com currently uses 30 days.

**Commitments**

For the duration of the HTTPSSVC or Alt-Svc lifetime, the server operator has made a commitment to the client. ESNI is a soft commitment that the server understands this ESNI key and a hard commitment that the server is colocated with the public name. The first lower-bounds key lifetime and rotation on the server. There is a recovery mechanism, but it is expensive, so this is a soft commitment. The second is roughly a commitment to use a particular hosting provider. ESNI's retry mechanism requires the public name, so this is a hard commitment. Breaking this will knock out your site.

(Note Alt-Svc without ESNI was *not* a hosting provider commitment. A provider-specific Alt-Svc may fail if the site changes providers, but the client could still connect without Alt-Svc. ESNI must take this fallback away to prevent network downgrade.)

HTTPSSVC and Alt-Svc commitment timescales are qualitatively different. Saying ESNI servers must advertise in both, as implied by the spec today, means servers must incur a long-lived hosting provider commitment to deploy ESNI at all. (Or send no Alt-Svc headers and lose QUIC on non-HTTPSSVC clients.) It also means ESNI keys must be long-lived, which makes them more sensitive.

**Proposed fix**

Given the above, I don't see how allowing Alt-Svc to override HTTPSSVC is tenable. That suggests changing the spec so a client that makes HTTPSSVC queries makes them even if Alt-Svc is available. If it gets an HTTPSSVC record, it ignores Alt-Svc and uses those instead. Otherwise, it may freely use Alt-Svc.

This is fussy because Alt-Svc itself allows replacing the origin hostname with an alternate name. Clients would likely want to query the alternate's A/AAAA records, the origin's A/AAAA records, and the origin/s HTTPSSVC records in parallel. However, the alternate may leak ESNI, so the alternate connection must wait for whether HTTPSSVC aborts it before proceeding past that query.

That adds *even more* complexity to the prospect of actually implementing remote Alt-Svc. Personally, I think all these name indirections are seeming more and more like a mistake and questionably worthwhile.

**Whither ESNI in Alt-Svc?**

With the above, it is no longer strictly necessary to allocate a way to spell ESNI in Alt-Svc. I don't know whether we still want to. This issue means ESNI in Alt-Svc is very different from ESNI in HTTPSSVC. At minimum, we must clearly call out the implications of the longer lifetime in the spec. We could decide this is not worth the trouble. On the other hand, it's likely a number of clients won't make HTTPSSVC queries for some time, and perhaps those clients getting ESNI for the subset of servers willing to make a longer-term public name commitment is worthwhile.

**Parting thought**

I think a lesson here is we cannot completely abstract ESNI from its delivery mechanism. Pulling ESNI into HTTPSSVC is reasonable so we only have one record to query, but HTTPSSVC's decisions still have implications for ESNI. (@chris-wood, I dunno if you watch this repo, so CC'ing you in here explicitly.)

**Comment by davidben** [comment]
The proposed fix does have an inverse consequence which is that Alt-Svc is *more* trustworthy than HTTPSSVC because it came from the origin. That makes overriding it odd. For everything currently delivered over Alt-Svc and HTTPSSVC, this is fine, but we are effectively constraining Alt-Svc to never depend on this trustworthiness in the future. (I.e. it can only be used for DNS-like things.)

This is hopefully fine because we can always make a new header (and probably should given all of the mistakes in Alt-Svc), but this is an implication to keep in mind.

**Comment by bemasc** [comment]
Thanks for thinking through all these problems, @davidben.  I've posted a PR (#106) that provides a fully detailed set of requirements for the interaction here.  I think this PR gets us to a solution that makes sensible configurations fully downgrade-resistant, avoids pathologies when transitioning between hosting providers, and also covers all the use cases we're aware of.  It continues to allow the behavior you're describing (HTTPSSVC having precedence).

I admit it's complex, but that's mostly because it's trying to describe a variety of potential client implementations.  I think the behavior of any specific client isn't really more complicated, just better specified.

**Comment by davidben** [comment]
Using HTTPSSVC to repair Alt-Svc errors rather than checking them ahead of time is an interesting idea, but it does not address the problem for clients that cannot make HTTPSSVC queries. Indeed the PR notes:

> If the client does not support HTTPSSVC, it MUST fail the connection to avoid enabling a downgrade attack.

That means a server still cannot put ESNI in Alt-Svc without long-term commitment about their hosting provider. While the PR does claim Alt-Svc is unnecessary:

> Origins MAY use esniconfig in Alt-Svc, HTTPSSVC, both, or neither

This is false. Saying it will break sporadically doesn't change the fact that it's broken. It also isn't sporadic: it will deterministicly break for all but the first connection. That does not count as deploying ESNI.

The section "Non-authoritative servers" needs to go in ESNI with full TLSWG analysis as it has security consequences for both the client and server and affects TLS itself. As to those security consequences, I do not think we should do this. Extracting an authenticated signal out of explicitly getting a bad certificate is *extremely* questionable. Certificate errors may arise from all kinds of reasons. Overloading that would be a problem for a client implementation.

Additionally, the text mandates particular behavior on unknown names. Given the subtle problems that come up when HTTP servers accept unknown Host headers, I don't think we should introduce such requirements, certainly not via an implicit signal like this.

This also still assumes the previous hosting provider behaves in a particular way.

Finally, this is far too much complexity on the client. Please use the *much* simpler solution outlined above.

**Comment by bemasc** [comment]
A trivial "HTTPSSVC first" implementation would forfeit resistance to downgrade attacks by DNS intermediaries, which I think is valuable.  However, I appreciate that the authenticated fallback proposal requires some unusual layer-crossing, and represents some kind of shift in the long-term commitment (although you already rely on your old host not to serve 404s with your cert).

I've updated #106 with a much shorter proposal that no longer discusses authenticated fallback.  It allows a simple client like you seem to be advocating for (HTTPSSVC first, no esniconfig in Alt-Svc), but also specifies how to do Alt-Svc in esniconfig safely, with a big warning label.

**Comment by chris-wood** [comment]
I’d opt for @davidben’s proposal, perhaps even with the additional condition that Alt-Svc not be a permissible vehicle for ESNIConfig values. The fact that these hang around for weeks at a time seems to encourage poor key management hygiene.

**Comment by bemasc** [comment]
I have concerns about any approach where HTTPSSVC without ESNI takes precedence over Alt-Svc ESNI.  That arrangement creates an incentive for hostile DNS servers to synthesize fake HTTPSSVC records, in order to override the ESNIConfig received in Alt-Svc.

We can certainly abandon ESNIConfig in Alt-Svc, but I think the loss is significant, and not to be done lightly.  Insofar as HTTPSSVC depends on encrypted DNS (as major clients seem likely to require), the vast majority of users will not be able to use it for the foreseeable future.  Alt-Svc ESNI is the only near-term path to broad deployment, and broad deployment is important to avoid ossification.

Note that Alt-Svc ESNI is not HPKP.  The ESNI fallback mechanism allows key rotation, so long as the fallback name doesn't change.  Yes, for good performance you shouldn't set your Alt-Svc lifetime longer than your key rotation period, but occasional clock skews or unscheduled rotations won't break anything.

As for lifetimes, ESNI keys are substantially less sensitive than TLS keys, so requiring them to be rotated ultra-frequently seems like a low priority.  Typical TLS certificates last for 90 days.  A similar or shorter lifetime for ESNI keys would easily be compatible with Alt-Svc ESNI.

**Comment by enygren** [comment]
Would it make sense to consider the Alt-Svc TTL as part of priorities here?  (ie, some platforms use Alt-Svc TTLs in the range of an hour.)  

Note that Alt-Svc and ESNI don't work well together unless clients honor alternate server DNS names separate from the hostname.  (ie, otherwise the same multi-server/multi-CDN issue shows up for Alt-Svc ESN.)  And in that model it is saner anyways to bound the TTL to something reasonable.  But reasonably short TTLs mostly make sense if clients honor Alt-Svc going to alternate servernames anyways as you want some degree of agility on these.

(One good use-case would be opportunistically supporting ESNI with QUIC prior to being ready to support it in TLS and before clients support doing HTTPSSVC lookups.)

**Comment by davidben** [comment]
> I have concerns about any approach where HTTPSSVC without ESNI takes precedence over Alt-Svc ESNI. That arrangement creates an incentive for hostile DNS servers to synthesize fake HTTPSSVC records, in order to override the ESNIConfig received in Alt-Svc.

It's certainly true that tying ESNI to DNS has consequences given a hostile DNS server. Both that and the dependency on new DNS records (which effectively depends on DoH) are indeed disappointing. But both properties have been true throughout this iteration of ESNI. We switched to the HTTPSSVC record on the assumption that it was equivalent to the original ESNI record. The Alt-Svc flaw breaks that assumption.

If we want to solve the hostile DNS server problem, we would need to get rid of all Alt-Svc failure paths on the client. The hostile DNS server also controls A/AAAA records, so it could blackhole `alt-svc-entrypoint.cdn.example` and cause all Alt-Svc routes to fail. (If we punt that problem and try only to solve the broad deployment problem, an Alt-Svc to HTTPSSVC fallback is tenable, though it alone doesn't solve the name pinning problem.)

Unless we can make ESNI in Alt-Svc truly equivalent to HTTPSSVC (unlikely given the TTL), it *must* be possible for servers to use HTTPSSVC with ESNI, use Alt-Svc without ESNI, and *still* get ESNI support for clients which support both. That means HTTPSSVC *must* be able to override Alt-Svc.

If we can come up with a way to salvage ESNI in Alt-Svc, we can always add some Alt-Svc directive that means this configuration cannot be overridden by HTTPSSVC. (This, of course, would still need to meet all the usual deployability and availability requirements.) We'd also still need to backport all of the ALPN-related fixes or general Alt-Svc support is dead in the water. All of the issues raised in #73 apply to Alt-Svc once ESNI shows up.

> As for lifetimes, ESNI keys are substantially less sensitive than TLS keys, so requiring them to be rotated ultra-frequently seems like a low priority. Typical TLS certificates last for 90 days. A similar or shorter lifetime for ESNI keys would easily be compatible with Alt-Svc ESNI.

I agree they're less sensitive. I'm more concerned about the name pinning than the key rotation.

A footnote on the equivalence to TLS keys however: 90 days is still too long for TLS certificates (the world needs to move to issuance automation). Also, as long-lived keys are fundamentally more sensitive than short-lived keys, server deployments apply extra protections to their long-lived keys like keeping them off of serving frontends. If ESNI keys, due to their lifetime, end up warranting such protections, it would double the cost of such measures.

> But reasonably short TTLs mostly make sense if clients honor Alt-Svc going to alternate servernames anyways as you want some degree of agility on these.

Certainly switching the names increases the desire for shorter TTLs, but that doesn't change the nature of Alt-Svc. Alt-Svc fundamentally only applies to *subsequent* connections, so the lifetime must extend to the *next* time you visit the site. (The current visit already has a preexisting connection. Even in clients willing to bear the complexity of cycling to new connection instantly, this is not useful for ossification purposes because blocking the alternate has no availability consequences while the main connection lives.)

The requirements on large enough Alt-Svc TTLs get stronger with ESNI. If you only use Alt-Svc as a routing optimization, perhaps you're fine with low Alt-Svc coverage as it's just an optimization. An ESNI-bearing Alt-Svc with small TTL barely applies.

> (One good use-case would be opportunistically supporting ESNI with QUIC prior to being ready to support it in TLS and before clients support doing HTTPSSVC lookups.)

This use case doesn't make sense. Remember that clients will always try TCP. We often say "fallback", but this is misleading. One of the failure modes is blackholed packets, so the two are actually run in parallel, so doing this will deterministically leak the SNI.

Indeed the one of the requirements on any fix to the many Alt-Svc ALPN mistakes is to *prevent* this, so that servers do not accidentally deploy a QUIC-only ESNI. Out-of-band signal => fallback for robustness => failing to provide equivalent security is a server configuration error.

**Comment by bemasc** [comment]
I believe #106 specified the behavior you're requesting here, so I'm closing this (as fixed).
